apis/mutations/unversioned/externaldata.go

```diff
@@ -0,0 +1,167 @@
+	// spec will be extracted to the external data provider as the data source.
+	DataSourceValueAtLocation DataSource = "ValueAtLocation"
+
+	// Username indicates that the username of the admission request will
+	// be extracted to the external data provider as the data source.
+	DataSourceUsername DataSource = "Username"
+)
+
+// FailurePolicy is the type of the failure policy to use for the external data.
+// +kubebuilder:validation:Enum=UseDefault;Ignore;Fail
+type FailurePolicy string
+
+const (
+	// UseDefault indicates that the default value of the external
+	// data provider will be used.
+	FailurePolicyUseDefault FailurePolicy = "UseDefault"
+
+	// Ignore indicates that the mutation will be ignored if the external
+	// data provider fails.
+	FailurePolicyIgnore FailurePolicy = "Ignore"
+
+	// Fail indicates that the mutation will be failed if the external
+	// data provider fails.
+	FailurePolicyFail FailurePolicy = "Fail"
+)
+
+// ExternalData describes the external data source to use for the mutation.
+type ExternalData struct {
+	// Provider is the name of the external data provider.
+	// +kubebuilder:validation:Required
+	Provider string `json:"provider,omitempty"`
+
+	// DataSource specifies where to extract the data that will be sent
+	// to the external data provider as parameters.
+	// +kubebuilder:default="ValueAtLocation"
+	DataSource DataSource `json:"dataSource,omitempty"`
+
+	// FailurePolicy specifies the policy to apply when the external data
+	// provider returns an error.
+	// +kubebuilder:default="Fail"
+	FailurePolicy FailurePolicy `json:"failurePolicy,omitempty"`
+
+	// Default specifies the default value to use when the external data
+	// provider returns an error and the failure policy is set to "UseDefault".
+	// +kubebuilder:validation:Schemaless
+	// +kubebuilder:validation:XPreserveUnknownFields
+	Default *types.Anything `json:"default,omitempty"`
+}
+
+func (e *ExternalData) Validate() error {
+	if !*externaldata.ExternalDataEnabled {
+		return ErrExternalDataFeatureFlagNotEnabled
+	}
+	if e.FailurePolicy == FailurePolicyUseDefault && e.Default == nil {
+		return ErrExternalDataFeatureNoDefault
+	}
+
+	return nil
+}
+
+// GetValue sends a request to the external data provider to get the value.
+// It returns a map of the outbound data and its response as key-value pairs.
+func (e *ExternalData) GetValue(param *types.GetValueParameter) (map[types.ProviderOutboundData]interface{}, error) {
+	sendRequest := types.DefaultSendRequestToExternalDataProvider
+	if param.SendRequestToExternalDataProvider != nil {
+		sendRequest = param.SendRequestToExternalDataProvider
+	}
+
+	provider, err := param.ProviderCache.Get(e.Provider)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get external data provider %s from cache: %w", e.Provider, err)
+	}
+
+	var keys []string
+	switch e.DataSource {
+	case DataSourceValueAtLocation:
+		valueAtLocation, err := param.Path.GetValueAtLocation(param.Mutable.Object.Object)
```

Comment:
By the time we're calling `GetValue()`, we are already at the object's location, could we just pass the current value directly? That would avoid duplicating the path traversal code.

Also, note that `location` can be ambiguous. E.g.: what's the location of `spec.containers[name: *].image` if we have a pod running 2 containers? We'd need to know the specific field the mutation logic is referring to to answer that.

Reply:
Take a look at my comment as to a potential execution flow (megacomment on `GetValueAtLocation()`), I think it's also relevant to this thread.